### PR TITLE
Basic BenchmarkDotNet support for Xamarin

### PR DESCRIPTION
**Pull request by jonathanpeppers**
Fixes: https://github.com/dotnet/BenchmarkDotNet/issues/1343

This is basically an "MVP" of Xamarin support with BenchmarkDotNet.

At a high level:

1. Simple benchmarks work with default config settings.

2. You can write a Xamarin.Forms app that runs benchmarks and displays
   the results. I added a very basic sample of this.

Some of the details to get this working:

* `ConsoleExitHandler.CancelKeypress` throws
  `PlatformNotSupportedException` on Xamarin platforms. Just ignore
  this exception.

* Multi-target `BenchmarkDotNet.Samples` to include `netstandard2.0`.
  This allows me to reference the benchmarks from the Xamarin.Forms
  shared code. I had to exclude some Windows-specific benchmarks.

* Don't probe for `mono` binary for `InProcess`. This was breaking
  `InProcess` for Xamarin, in general. It's not going to find a `mono`
  binary that doesn't exist.

* Detect Xamarin.Android / Xamarin.iOS. I used the most common
  `Type.GetType` lookup to detect each platform: `Java.Lang.Object`
  and `Foundation.NSObject`.

* Fix for `Directory.CurrentDirectory()` as `ArtifactPath`. The
  current directory on Android is not writeable. Looking for `/`,
  seemed like a simple fix for Android.

* Default to `InProcess` for Xamarin platforms. Since we can't find a
  `mono` binary, we have to run in-process. Xamarin.iOS also shouldn't
  use `System.Reflection.Emit` until the interpreter is more stable.
  Users of BDN can experiment with the interpreter by configuring
  their benchmarks manually.

This is the first I've looked at this repo, so suggestions welcome! I don't
know this codebase.

The bulk of the new files here is what you get from a new Xamarin.Forms
template in Visual Studio. We can trim that down if needed. Thanks!

Screen Shot 2020-01-27 at 8 47 32 AM

### CODE REVIEW

**Comment by jonathanpeppers** [comment]
I'll look into the CI failures tomorrow:
```
"C:\projects\benchmarkdotnet\.dotnet\sdk\2.1.503\Xamarin\Android\Xamarin.Android.CSharp.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk.
    135 Warning(s)
    1 Error(s)
Time Elapsed 00:00:58.99
```
I think AppVeyor can use 2019 image. I also have a build warning in the Forms project I can fix.

**Comment by jonathanpeppers** [comment]
Questions:
1. The Xamarin projects won't be able to build with `dotnet build` until we ship .NET 5. Should I just remove them from the main SLN file for now?
2. Is there a doc I should update to mention how to use with Xamarin?

**Comment by jonathanpeppers** [comment]
> 1. The Xamarin projects won't be able to build with dotnet build until we ship .NET 5. Should I just remove them from the main SLN file for now?
> 2. Is there a doc I should update to mention how to use with Xamarin?

@adamsitnik or @AndreyAkinshin thoughts about this, thanks!

**Comment by adambarath** [comment]
Hi! Any updates on this? BR

**Comment by jonathanpeppers** [comment]
@adambarath I never heard anything back on this.

I'll use my best judgement to make this "mergeable" and maybe we'll hear back by then?

File: src/BenchmarkDotNet/Toolchains/ToolchainExtensions.cs [hunk 1]
```diff
@@ -39,6 +41,10 @@ internal static IToolchain GetToolchain(this Runtime runtime, Descriptor descrip
                     return RoslynToolchain.Instance;
 
                 case MonoRuntime mono:
+                    if (mono.IsXamarinAndroid)
+                        return InProcessEmitToolchain.Instance;
+                    if (mono.IsXamariniOS)
+                        return InProcessNoEmitToolchain.Instance;
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Toolchains/ToolchainExtensions.cs:47]: @ig-sinicyn nice use case for the no emit toolchain ;)

File: samples/BenchmarkDotNet.Samples.Forms/MainPage.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+			{
+				var logger = new AccumulationLogger();
+				await Task.Run(() =>
+				{
+					var summary = BenchmarkRunner.Run<IntroBasic>();
+					MarkdownExporter.Console.ExportToLog(summary, logger);
+					foreach (var benchmark in summary.BenchmarksCases)
+					{
+						ConclusionHelper.Print(logger, benchmark.Config.GetCompositeAnalyser().Analyse(summary).Distinct().ToList());
+					}
```
Comment (adamsitnik) [inline samples/BenchmarkDotNet.Samples.Forms/MainPage.xaml.cs:32]: in most of the cases, all benchmark instances are going to share the same instance of `Config`. To avoid duplicates we should aggregate the conclusions for all of them and then call distinct

```suggestion
        ConclusionHelper.Print(logger, 
            summary.BenchmarksCases
                   .SelectMany(benchmark => benchmark.Config.GetCompositeAnalyser().Analyse(summary))
                   .Distinct()
                   .ToList()
       );
```

File: samples/BenchmarkDotNet.Samples.iOS/Entitlements.plist [hunk 3]
```diff
@@ -0,0 +1,7 @@
+﻿<?xml version="1.0" encoding="UTF-8" ?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+</dict>
+</plist>
+
```
Comment (adamsitnik) [inline samples/BenchmarkDotNet.Samples.iOS/Entitlements.plist:7]: nit: empty line
```suggestion
```

File: samples/BenchmarkDotNet.Samples.Forms/BenchmarkDotNet.Samples.Forms.csproj [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+  <ItemGroup>
+    <PackageReference Include="System.Collections.Immutable" Version="1.7.0" />
+    <PackageReference Include="Xamarin.Forms" Version="4.5.0.530" />
+    <PackageReference Include="Xamarin.Essentials" Version="1.5.2" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\src\BenchmarkDotNet\BenchmarkDotNet.csproj" />
+    <ProjectReference Include="..\BenchmarkDotNet.Samples\BenchmarkDotNet.Samples.csproj" />
```
Comment (adamsitnik) [inline samples/BenchmarkDotNet.Samples.Forms/BenchmarkDotNet.Samples.Forms.csproj:16]: is the dependency on `BenchmarkDotNet.Samples` necessary here? Could we remove it? If so, could you please also remove it from the `BenchmarkDotNet.Xamarin` solution and revert changes applied to `BenchmarkDotNet.Samples.csproj`?

```suggestion
```

File: samples/BenchmarkDotNet.Samples/BenchmarkDotNet.Samples.csproj [hunk 5]
```diff
@@ -23,6 +23,11 @@
...
     <ProjectReference Include="..\..\src\BenchmarkDotNet\BenchmarkDotNet.csproj" />
-    <ProjectReference Include="..\..\src\BenchmarkDotNet.Diagnostics.Windows\BenchmarkDotNet.Diagnostics.Windows.csproj" />
+    <ProjectReference Condition=" '$(TargetFramework)' != 'netstandard2.0' " Include="..\..\src\BenchmarkDotNet.Diagnostics.Windows\BenchmarkDotNet.Diagnostics.Windows.csproj" />
+  </ItemGroup>
+  <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0' ">
+    <Compile Remove="IntroInliningDiagnoser.cs" />
+    <Compile Remove="IntroNativeMemory.cs" />
+    <Compile Remove="IntroTailcall.cs" />
   </ItemGroup>
 </Project>
```
Comment (adamsitnik) [inline samples/BenchmarkDotNet.Samples/BenchmarkDotNet.Samples.csproj:33]: as explained above: if the `Form` project is not really using the `Samples` project, we can remove the dependency and revert the changes  applied to this app

File: samples/BenchmarkDotNet.Samples.Forms/BenchmarkDotNet.Samples.Forms.csproj (hunk 4 above)
Comment (jonathanpeppers) [reply samples/BenchmarkDotNet.Samples.Forms/BenchmarkDotNet.Samples.Forms.csproj]: I just added a single link to this instead, which should be fine:
```xml
<Compile Include="..\BenchmarkDotNet.Samples\IntroBasic.cs" Link="Benchmarks\IntroBasic.cs" />
```

Comment (adamsitnik) [reply samples/BenchmarkDotNet.Samples.Forms/BenchmarkDotNet.Samples.Forms.csproj]: @jonathanpeppers great, thank you!

**Comment by wcoder** [comment]
It looks like the documentation hasn't been deployed yet
`https://benchmarkdotnet.org/articles/samples/IntroXamarin.html` ?
